spec/Candid.md

```diff
@@ -864,142 +864,213 @@ service { <methtype>;* } <: service { <methtype'>;* }
-It *optimistically* tries to decode an option value
-and succceeds if there would have been a valid type `<datatype''>` for the input value that is a subtype of both types.
-(As formulated, the rule would be non-deterministic in the choice of `<datatype''>`, but the intention is to pick the largest type that makes `f y` succeed if possible. We take the liberty to hand-wave over the formulation of this detail here.)
 
-The effect of this rule is that decoders do not actually need to perform separate subtype checks during deserialisation.
-Instead, they can simply try to deserialise, and if they encounter a mismatch abort to the innermost option type, returning `null`.
+An optional value decodes at an option type, if the value decodes at the constituent type:
+```
+<v> :? <t> ~> <v'>
+-----------------------
+opt <v> :? opt <t> ~> opt <v'>
+```
 
-*Note:* The working assumption is that the type describing an incoming value ought to be *principal*, i.e., the most precise type assignable to the value (in a sense that could be made precise).
-In that case, `<datatype''>` equals `<datatype>` and deserialization succeeds exactly if the types match.
+If an optional value _fails_ to decode at an optional type, the result is `null`, not failure:
+```
+<v> !: <t>
+-----------------------
+opt <v> :? opt <t> ~> null
+```
 
-However, in practice it would be costly to enforce the requirement that all type descriptions in a serialised value are principal, for both encoder (who would need to compute the principal type) and decoders (who would need to check it).
-For example, to be principal, an empty vector would need to have type `vec empty`, `null` could only have type `null`, and a variant would only be allowed to include tags that actually occur in the value.
-Consequently, the semantics allows for the possibility that the encoder ascribes a less specific type `<datatype>` with some redundant information, and allows the decoder to ignore redundant information (by going to `<datatype''>`) like the element type of empty arrays, of null options, or unused variant tags.
+Decoding a non-null, non-optional and non-reserved value at an option type treats it as an optional value:
+```
+<v> ≠ null
+<v> ≠ (null : reserved)
+<v> ≠ opt _
+opt <v> :? opt <t> ~> <v'>
+-----------------------
+<v> :? opt <t> ~> <v'>
+```
 
 
 #### Records
 
+Only records decode at record type. Missing fields of option type turn into `null`.
+
+In the following rule, the `<nat1>*` field names are those present in both the value and the type, the `<nat2>*` field names only those in the value, and `<nat3>*` are those only in the type.
+```
+<v1> :? <t1> ~> <v1'>
+--------------------------------------------------------------------------------------------------------------------------------------------
+record { <nat1> = <v1>;* <nat2> = <v2>;* } :? record {  <nat1> = <t1>;* <nat3> = opt <t2>;* } ~> record { <nat1> = <v1'>;* <nat3> = null;* }
```

Comment:
This needs a tweak in light of https://github.com/dfinity/candid/issues/131 to allow for `reserved`

Reply:
Will have a look.